src/Microsoft.NetCore.Analyzers/Core/Security/DoNotReferSelfInSerializableClass.cs

```diff
@@ -0,0 +1,255 @@
+                    var inDegree = new ConcurrentDictionary<ISymbol, int>();
+
+                    // It Keeps the out degree of every vertex in the forwardGraph, which is corresponding to the in Degree of the vertex in invertedGraph.
+                    var outDegree = new ConcurrentDictionary<ISymbol, int>();
+
+                    compilationStartAnalysisContext.RegisterSymbolAction(
+                        (SymbolAnalysisContext symbolAnalysisContext) =>
+                        {
+                            DrawGraph(symbolAnalysisContext.Symbol);
+                        }, SymbolKind.NamedType);
+
+                    compilationStartAnalysisContext.RegisterCompilationEndAction(
+                        (CompilationAnalysisContext compilationAnalysisContext) =>
+                        {
+                            ModifyDegree(inDegree, forwardGraph);
+                            ModifyDegree(outDegree, invertedGraph);
+
+                            // If the degree of a vertex is greater than 0 both in the forward graph and inverted graph after topological sorting,
+                            // the vertex must belong to a loop.
+                            var leftVertices = inDegree.Where(s => s.Value > 0).Select(s => s.Key).ToImmutableHashSet();
+                            var invertedLeftVertices = outDegree.Where(s => s.Value > 0).Select(s => s.Key).ToImmutableHashSet();
+                            var verticesInLoop = leftVertices.Intersect(invertedLeftVertices);
+
+                            foreach (var vertex in verticesInLoop)
+                            {
+                                if (vertex is IFieldSymbol fieldInLoop)
+                                {
+                                    var associatedSymbol = fieldInLoop.AssociatedSymbol;
+                                    compilationAnalysisContext.ReportDiagnostic(
+                                        fieldInLoop.CreateDiagnostic(
+                                            Rule,
+                                            associatedSymbol == null ? vertex.Name : associatedSymbol.Name));
+                                }
+                            }
+                        });
+
+                    /// <summary>
+                    /// Traverse from point to its descendants, save the information into a directed graph.
+                    /// </summary>
+                    /// <param name="point">The initial point</param>
+                    void DrawGraph(ISymbol point)
+                    {
+                        // If the point has been visited, return;
+                        // otherwise, add it to the graph and mark it as visited.
+                        if (!AddPointToBothGraphs(point))
+                        {
+                            return;
+                        }
+
+                        if (point is INamedTypeSymbol namedTypePoint)
```

Comment:
Are the series of `if` statements here mutually exclusive?  If so, could do `else if` to make that clear.